HoconConfigParser does not support null values

Parsing a JSON config file with a `null` value results in a NPE:

```
Exception in thread "main" io.helidon.config.spi.ConfigParserException: Cannot read from source: null
	at io.helidon.config.hocon.HoconConfigParser.parse(HoconConfigParser.java:139)
	at io.helidon.config.ConfigSourceRuntimeImpl$ParsableConfigSourceReloader.lambda$get$0(ConfigSourceRuntimeImpl.java:477)
	at java.base/java.util.Optional.map(Optional.java:265)
	at io.helidon.config.ConfigSourceRuntimeImpl$ParsableConfigSourceReloader.get(ConfigSourceRuntimeImpl.java:460)
	at io.helidon.config.ConfigSourceRuntimeImpl$ParsableConfigSourceReloader.get(ConfigSourceRuntimeImpl.java:443)
	at java.base/java.util.Optional.orElseGet(Optional.java:369)
	at io.helidon.config.ConfigSourceRuntimeImpl.initialLoad(ConfigSourceRuntimeImpl.java:218)
	at io.helidon.config.ConfigSourceRuntimeImpl.load(ConfigSourceRuntimeImpl.java:173)
	at io.helidon.config.ConfigSourcesRuntime.load(ConfigSourcesRuntime.java:140)
	at io.helidon.config.ProviderImpl.newConfig(ProviderImpl.java:87)
	at io.helidon.config.BuilderImpl.build(BuilderImpl.java:333)
	at io.helidon.config.BuilderImpl.build(BuilderImpl.java:51)
	at io.helidon.config.Config.create(Config.java:339)
	at io.helidon.config.examples.basics.Main.main(Main.java:40)
Caused by: java.lang.NullPointerException
	at io.helidon.config.hocon.HoconConfigParser.lambda$fromConfig$0(HoconConfigParser.java:152)
	at java.base/java.util.Map.forEach(Map.java:661)
	at io.helidon.config.hocon.HoconConfigParser.fromConfig(HoconConfigParser.java:145)
	at io.helidon.config.hocon.HoconConfigParser.lambda$fromList$1(HoconConfigParser.java:164)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at io.helidon.config.hocon.HoconConfigParser.fromList(HoconConfigParser.java:160)
	at io.helidon.config.hocon.HoconConfigParser.lambda$fromConfig$0(HoconConfigParser.java:148)
	at java.base/java.util.Map.forEach(Map.java:661)
	at io.helidon.config.hocon.HoconConfigParser.fromConfig(HoconConfigParser.java:145)
	at io.helidon.config.hocon.HoconConfigParser.lambda$fromConfig$0(HoconConfigParser.java:150)
	at java.base/java.util.Map.forEach(Map.java:661)
	at io.helidon.config.hocon.HoconConfigParser.fromConfig(HoconConfigParser.java:145)
	at io.helidon.config.hocon.HoconConfigParser.lambda$fromConfig$0(HoconConfigParser.java:150)
	at java.base/java.util.Map.forEach(Map.java:661)
	at io.helidon.config.hocon.HoconConfigParser.fromConfig(HoconConfigParser.java:145)
	at io.helidon.config.hocon.HoconConfigParser.parse(HoconConfigParser.java:135)
	... 13 more

```

The culprit is this statement: `value.unwrapped().toString()` ; `value.unwrapped()` returns `null`.

Seems like `YamlConfigParser` handles `null` as a value that holds an empty string:

```java
String strValue = v == null ? "" : v.toString();
builder.addValue(strKey, strValue);
```

We probably want to do the same here to remain consistent (even though it seems odd).

--- a/config/hocon/src/main/java/io/helidon/config/hocon/HoconConfigParser.java
+++ b/config/hocon/src/main/java/io/helidon/config/hocon/HoconConfigParser.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -149,7 +149,12 @@ public class HoconConfigParser implements ConfigParser {
             } else if (value instanceof ConfigObject) {
                 builder.addObject(key, fromConfig((ConfigObject) value));
             } else {
-                builder.addValue(key, value.unwrapped().toString());
+                Object unwrapped = value.unwrapped();
+                if (unwrapped == null) {
+                    builder.addValue(key, "");
+                } else {
+                    builder.addValue(key, String.valueOf(unwrapped));
+                }
             }
         });
         return builder.build();

--- a/config/hocon/src/main/java/io/helidon/config/hocon/HoconConfigParserBuilder.java
+++ b/config/hocon/src/main/java/io/helidon/config/hocon/HoconConfigParserBuilder.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2017, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -70,6 +70,7 @@ public final class HoconConfigParserBuilder implements Builder<ConfigParser> {
      *
      * @return new instance of HOCON ConfigParser.
      */
+    @Override
     public HoconConfigParser build() {
         return new HoconConfigParser(resolvingEnabled, resolveOptions);
     }

--- a/config/hocon/src/test/java/io/helidon/config/hocon/HoconConfigParserTest.java
+++ b/config/hocon/src/test/java/io/helidon/config/hocon/HoconConfigParserTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2017, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -207,7 +207,7 @@ public class HoconConfigParserTest {
 
         Config config = Config
                 .builder(ConfigSources.create(JSON, HoconConfigParser.MEDIA_TYPE_APPLICATION_JSON))
-                .addParser(new HoconConfigParser())
+                .addParser(HoconConfigParser.create())
                 .disableEnvironmentVariablesSource()
                 .disableSystemPropertiesSource()
                 .disableParserServices()
@@ -252,7 +252,7 @@ public class HoconConfigParserTest {
 
     @Test
     public void testGetSupportedMediaTypes() {
-        HoconConfigParser parser = new HoconConfigParser();
+        HoconConfigParser parser = HoconConfigParser.create();
 
         assertThat(parser.supportedMediaTypes(), is(not(empty())));
     }
@@ -261,7 +261,7 @@ public class HoconConfigParserTest {
     public void testCustomTypeMapping() {
         Config config = Config
                 .builder(ConfigSources.create(AppType.DEF, HoconConfigParser.MEDIA_TYPE_APPLICATION_JSON))
-                .addParser(new HoconConfigParser())
+                .addParser(HoconConfigParser.create())
                 .addMapper(AppType.class, new AppTypeMapper())
                 .disableEnvironmentVariablesSource()
                 .disableSystemPropertiesSource()
@@ -275,6 +275,23 @@ public class HoconConfigParserTest {
         assertThat("page-size", app.getPageSize(), is(AppType.PAGE_SIZE));
         assertThat("basic-range", app.getBasicRange(), is(AppType.BASIC_RANGE));
 
+    }
+
+    @Test
+    void testParserFromJson() {
+        Config config = Config.builder()
+                .disableSystemPropertiesSource()
+                .disableEnvironmentVariablesSource()
+                .disableParserServices()
+                .addParser(HoconConfigParser.create())
+                .addSource(ConfigSources.classpath("config.json"))
+                .build();
+
+        Optional<String> property = config.get("oracle.com").asString().asOptional();
+        assertThat(property, is(Optional.of("1")));
+
+        property = config.get("nulls.null").asString().asOptional();
+        assertThat(property, is(Optional.of("")));
     }
 
     //
@@ -317,11 +334,11 @@ public class HoconConfigParserTest {
                 + "  storagePassphrase = \"${AES=thisIsEncriptedPassphrase}\""
                 + "}";
 
-        private String greeting;
-        private String name;
-        private int pageSize;
-        private List<Integer> basicRange;
-        private String storagePassphrase;
+        private final String greeting;
+        private final String name;
+        private final int pageSize;
+        private final List<Integer> basicRange;
+        private final String storagePassphrase;
 
         public AppType(
                 String name,
@@ -365,15 +382,14 @@ public class HoconConfigParserTest {
 
         @Override
         public AppType apply(Config config) throws ConfigMappingException, MissingValueException {
-            AppType app = new AppType(
+
+            return new AppType(
                     config.get("name").asString().get(),
                     config.get("greeting").asString().get(),
                     config.get("page-size").asInt().get(),
                     config.get("basic-range").asList(Integer.class).get(),
                     config.get("storagePassphrase").asString().get()
             );
-
-            return app;
         }
     }
 }

--- a/config/hocon/src/test/resources/config.json
+++ b/config/hocon/src/test/resources/config.json
@@ -6,5 +6,8 @@
     "oracle": {
         "com": "1",
         "cz": "2"
+    },
+    "nulls": {
+        "null": null
     }
 }

